libmultipath/public_api.c

```diff
@@ -0,0 +1,794 @@
+	struct dmmp_path *mp_path_new = NULL;
+	struct dmmp_path *mp_path_old = NULL;
+	uint32_t i = 0;
+	uint32_t j = 0;
+
+	if (mpath == NULL)
+		return NULL;
+
+	mpath_new = _dmmp_mpath_new();
+	if (mpath_new == NULL)
+		return NULL;
+
+	memcpy(mpath_new, mpath, sizeof(struct dmmp_mpath));
+	mpath_new->mp_pgs = NULL;
+	mpath_new->alias = NULL;
+	mpath_new->mp_pg_count = 0;
+
+	if (mpath->alias != NULL) {
+		mpath_new->alias = STRDUP(mpath->alias);
+		if (mpath_new->alias == NULL)
+			goto error_out;
+	}
+
+	if (mpath->mp_pg_count == 0)
+		return mpath_new;
+
+	mpath_new->mp_pgs = (struct dmmp_path_group **)
+		MALLOC(sizeof(struct dmmp_path_group *) * mpath->mp_pg_count);
+
+	if (mpath_new->mp_pgs == NULL)
+		goto error_out;
+
+	mpath_new->mp_pg_count = mpath->mp_pg_count;
+
+	for (i = 0; i < mpath->mp_pg_count; ++i) {
+		mp_pg_old = mpath->mp_pgs[i];
+		mp_pg_new = _dmmp_path_group_new();
+		if (mp_pg_new == NULL)
+			goto error_out;
+		mpath_new->mp_pgs[i] = mp_pg_new;
+
+		memcpy(mp_pg_new, mp_pg_old, sizeof(struct dmmp_path_group));
+		mp_pg_new->selector = NULL;
+		mp_pg_new->mp_paths = NULL;
+		mp_pg_new->mp_path_count = 0;
+
+		mp_pg_new->selector = STRDUP(mp_pg_old->selector);
+		if (mp_pg_new->selector == NULL)
+			goto error_out;
+		if (mp_pg_old->mp_path_count == 0)
+			continue;
+
+		mp_pg_new->mp_paths = (struct dmmp_path **)
+			MALLOC(sizeof(struct dmmp_path *) *
+			       mp_pg_old->mp_path_count);
+		if (mp_pg_new->mp_paths == NULL)
+			goto error_out;
+
+		mp_pg_new->mp_path_count = mp_pg_old->mp_path_count;
+
+		for (j = 0; j < mp_pg_old->mp_path_count; ++j) {
+			mp_path_old = mp_pg_old->mp_paths[j];
+			mp_path_new = _dmmp_path_new();
+			if (mp_path_new == NULL)
+				goto error_out;
+			mp_pg_new->mp_paths[j] = mp_path_new;
+
+			memcpy(mp_path_new, mp_path_old,
+			       sizeof(struct dmmp_path));
+		}
+	}
+
+	return mpath_new;
+
+error_out:
+	dmmp_mpath_free(mpath_new);
+	return NULL;
+}
+
+int dmmp_path_group_list_get(struct dmmp_mpath *mpath,
+			     struct dmmp_path_group ***mp_pgs,
+			     uint32_t *mp_pg_count)
+{
+	*mp_pgs = NULL;
+	*mp_pg_count = 0;
+	if (mpath != NULL)
+		*mp_pgs = mpath->mp_pgs;
+		*mp_pg_count = mpath->mp_pg_count;
+	return 0;
+}
+
+uint32_t dmmp_path_group_id_get(struct dmmp_path_group *mp_pg)
+{
+	if (mp_pg == NULL)
+		return 0;
+	return mp_pg->id;
+}
+
+uint32_t dmmp_path_group_priority_get(struct dmmp_path_group *mp_pg)
+{
+	if (mp_pg == NULL)
+		return 0;
+	return mp_pg->priority;
+}
+
+
+enum dmmp_path_group_status
+dmmp_path_group_status_get(struct dmmp_path_group *mp_pg)
+{
+	if (mp_pg == NULL)
+		return DMMP_PATH_GROUP_STATUS_UNDEF;
+	return mp_pg->status;
+}
+
+const char *dmmp_path_group_selector_get(struct dmmp_path_group *mp_pg)
+{
+	if (mp_pg == NULL)
+		return NULL;
+	return mp_pg->selector;
+
+}
+
+int dmmp_path_list_get(struct dmmp_path_group *mp_pg,
+		       struct dmmp_path ***mp_paths,
+		       uint32_t *mp_path_count)
+{
+	*mp_paths = NULL;
+	*mp_path_count = 0;
```

Comment:
Sorry, I tagged the wrong line with this comment.  I was commenting about the indentation of this...   you don't want to indent the 2nd line since it isn't conditional, right?


+	if (mp_pg != NULL)
+		*mp_paths = mp_pg->mp_paths,
+		*mp_path_count = mp_pg->mp_path_count;